llvm/lib/Target/NVPTX/SYCL/GlobalOffset.cpp

```diff
@@ -0,0 +1,368 @@
+        for (Use &U : CallToOld->args()) {
+          ImplicitOffsets.push_back(U);
+        }
+        ImplicitOffsets.push_back(ImplicitOffset);
+
+        // Replace call to other function (which now has a new parameter),
+        // with a call including the new parameter to that same function.
+        auto NewCaller = CallInst::Create(
+            /* Ty= */ CalleeWithImplicitParam->getFunctionType(),
+            /* Func= */ CalleeWithImplicitParam,
+            /* Args= */ ImplicitOffsets,
+            /* NameStr= */ Twine(),
+            /* InsertBefore= */ CallToOld);
+        NewCaller->setTailCallKind(CallToOld->getTailCallKind());
+        CallToOld->replaceAllUsesWith(NewCaller);
+
+        if (CallToOld->hasName()) {
+          NewCaller->takeName(CallToOld);
+        }
+      }
+
+      // Remove the caller now that it has been replaced.
+      CallToOld->eraseFromParent();
+
+      if (!AlreadyProcessed) {
+        // Process callers of the old function.
+        this->addImplicitParameterToCallers(M, Caller, NewFunc);
+
+        // Now that the old function is dead, delete it.
+        Caller->dropAllReferences();
+        Caller->eraseFromParent();
+      }
+    }
+  }
+
+  std::pair<Function *, Value *>
+  addOffsetArgumentToFunction(Module &M, Function *Func,
+                              Type *ImplicitArgumentType = nullptr,
+                              bool KeepOriginal = false) {
+    FunctionType *FuncTy = Func->getFunctionType();
+    const AttributeList &FuncAttrs = Func->getAttributes();
+    ImplicitArgumentType =
+        ImplicitArgumentType ? ImplicitArgumentType : ImplicitOffsetPtrType;
+
+    // Construct an argument list containing all of the previous arguments.
+    SmallVector<Type *, 8> Arguments;
+    SmallVector<AttributeSet, 8> ArgumentAttributes;
+
+    unsigned i = 0;
+    for (Function::arg_iterator FuncArg = Func->arg_begin(),
+                                FuncEnd = Func->arg_end();
+         FuncArg != FuncEnd; ++FuncArg, ++i) {
+      Arguments.push_back(FuncArg->getType());
+      ArgumentAttributes.push_back(FuncAttrs.getParamAttributes(i));
+    }
+
+    // Add the offset argument. Must be the same type as returned by
+    // `llvm.nvvm.implicit.offset`.
+
+    Arguments.push_back(ImplicitArgumentType);
+    ArgumentAttributes.push_back(AttributeSet());
+
+    // Build the new function.
+    AttributeList NAttrs =
+        AttributeList::get(Func->getContext(), FuncAttrs.getFnAttributes(),
+                           FuncAttrs.getRetAttributes(), ArgumentAttributes);
+    assert(!FuncTy->isVarArg() && "Variadic arguments prohibited in SYCL");
+    FunctionType *NewFuncTy = FunctionType::get(FuncTy->getReturnType(),
+                                                Arguments, FuncTy->isVarArg());
+
+    Function *NewFunc = Function::Create(NewFuncTy, Func->getLinkage(),
+                                         Func->getAddressSpace());
+
+    if (KeepOriginal) {
```

Comment:
I'm not sure if it will simplify the implementation, but instead of having two clones of entry points we might just rename the original entry point to `<entry_point_with_offset>` and add another small kernel w/o offset which sets offset to zero and calls `entry_point_with_offset>`.

```llvm
; Function Attrs: noinline
define weak_odr dso_local void @_ZTS14example_kernel() {
entry:
  %0 = alloca [3 x i32], align 4
  %1 = bitcast [3 x i32]* %0 to i8*
  call void @llvm.memset.p0i8.i64(i8* nonnull align 4 dereferenceable(12) %1, i8 0, i64 12, i1 false)
  %2 = getelementptr inbounds [3 x i32], [3 x i32]* %0, i32 0, i32 0
  %3 = call i64 @_ZTS14example_kernel_with_offset(i32* %2)
  ret void
}

define weak_odr dso_local void @_ZTS14example_kernel_with_offset([3 x i32]* byval([3 x i32]) %0) {
entry:
  %1 = bitcast [3 x i32]* %0 to i32*
  %2 = call i64 @_ZTS14other_function(i32* %1)
  ...
  ret void
}
```

Reply:
I don't think this will simplify the implementation much as you still need to rebuild everything that comes with the function. So instead of cloning you will have a few extra instructions inserted, so not much will change.

The main problem of doing this is it will break the assumption for the local memory handling pass that a kernel does not call another kernel (which doesn't happen in SYCL). We can order the pass to make sure this is safe, but this would create a secret handshake.